### PR TITLE
COMMERCE-7574 Batch-engine should allow skipping failed records

### PR DESCRIPTION
**Pull request by matijapetanjek**
/cc @javierdearcos

### CODE REVIEW

**Review by javierdearcos — COMMENTED** [review]
I'd love the approach and the abstraction of the ImportStrategy to allow modifying the behavior of the creation method.

One minor thing I would suggest is to rename the strategy object in Vulcan and ResourceImpls to something like CreationStrategy, as in the Vulcan domain we don't have any knowledge about import or export.

My main concern is that this would be a breaking change. Don't know if we could add a new create method with the new parameter with a default implementation and leave the old one calling to the new one with the default strategy to avoid this breaking change.

My concern is about clients that are already using the batch endpoints and would suffer problems when they regenerate the code

**Comment by matijapetanjek** [comment]
> I love the approach and the abstraction of the ImportStrategy to allow modifying the behavior of the creation method.
> 
> One minor thing I would suggest is to rename the strategy object in Vulcan and ResourceImpls to something like CreationStrategy, as in the Vulcan domain we don't have any knowledge about import or export.
> 
> My main concern is that this would be a breaking change. Don't know if we could add a new create method with the new parameter with a default implementation and leave the old one calling to the new one with the default strategy to avoid this breaking change.
> 
> My concern is about clients that are already using the batch endpoints and would suffer problems when they regenerate the code

Yes, you are

**Comment by matijapetanjek** [comment]
@javierdearcos Yes you're absolutely correct, I'll try to add a new create method instead of changing the existing one. The old method will call the new one with a DefaultCreationStrategy implementation which corresponds to FailOnError batch-engine strategy (same behavior as we have now).

I'll update this PR once I'm done with the changes.

Thank you!

**Comment by matijapetanjek** [comment]
@javierdearcos I've pushed the latest code which contains the changes we discussed on the slack. Please review, thank you!

File: modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/batch/engine/BatchStrategy.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.vulcan.batch.engine;
+
+import com.liferay.batch.engine.strategy.ImportStrategy;
```
Comment (javierdearcos) [inline modules/apps/portal-vulcan/portal-vulcan-api/src/main/java/com/liferay/portal/vulcan/batch/engine/BatchStrategy.java:17]: I think this import would force the module clients to add the batch-api dependency with the vulcan-api dependency.

I would prefer to have an interface BatchStrategy here without any references to batch in the API module and then have an Impl BatchStrategyImpl class in the impl module adapting the Import strategy.

Checking now, I realized that we already have a dependency in the portal-vulcan-api build.gradle file to the batch-engine-api but it is unnecessary and we should get rid of it

**Review by javierdearcos — CHANGES_REQUESTED** [review]
Please, see my comment about the BatchStrategy class.

Everything but this I consider is perfect. It was a great implementation and all fits really well and is very clean.

The only thing left to resolve is that with this generation change we are forcing the REST Builder clients to update Vulcan when they regenerate their code (because we are using new functions or classes that are not available in old packages).

This is something not related only to this PR and that will happen also with the PR that I'm preparing about providing the Batch metadata methods.

I need to think about how to avoid or alleviate this to not cause problems in REST Builder clients. I'll give it a thought and let you know

**Comment by matijapetanjek** [comment]
Thank you for the reviews Javier, I've pushed the requested changes and also removed batch-engine-api from the portal-vulcan-api dependencies. Running relevant tests just to check how changes impacted the rest of the headless infra.

**Comment by matijapetanjek** [comment]
ci:test:relevant

**Review by javierdearcos — APPROVED** [review]
Great!

Thank you very much for the effort and for attending to all the change requests.

Let's see the test results

**Comment by matijapetanjek** [comment]
ci:reevaluate:1277401_1970

**Comment by matijapetanjek** [comment]
ci:test:relevant

**Comment by matijapetanjek** [comment]
ci:test:sf

**Comment by javierdearcos** [comment]
Sent here https://github.com/liferay-frontend/liferay-portal/pull/1940

**Pull request closed without merging**